integrated_channels/canvas/client.py

```diff
@@ -51,9 +51,124 @@ def __init__(self, enterprise_configuration):
+            account_id=self.enterprise_configuration.canvas_account_id,
+            email_address=user_id
+        )
+        rsps = self.session.get(get_user_id_from_email_url)
+        rsps.raise_for_status
+        get_users_by_email_response = rsps.json()
+        try:
+            canvas_user_id = get_users_by_email_response[0]['id']
+        except (KeyError, IndexError):
+            # Trying to figure out how we should handle errors here- should we catch
+            # them in the in the transmitter? Should we return 400 but not raise?
+            # If we have multiple course completions to post, a raised exception here without
+            # anything else will prevent other transmissions to happen.
+            raise CanvasClientError(
+                "No Canvas user ID found associated with email: {}".format(learner_data['userID'])
+            )
+
+        # With the Canvas user ID, retrieve all courses for the user.
+        get_users_courses_url = '{canvas_base_url}/api/v1/users/{canvas_user_id}/courses'.format(
+            canvas_base_url=self.enterprise_configuration.canvas_base_url,
+            canvas_user_id=canvas_user_id
+        )
+        rsps = self.session.get(get_users_courses_url)
+        rsps.raise_for_status()
+        user_courses_response = rsps.json()
+
+        # Find the course who's integration ID matches the learner data course ID
+        # Is there a better identifier here?
+        course_id = None
+        for course in user_courses_response:
+            integration_id = course['integration_id']
+            if '+'.join(integration_id.split(":")[1].split("+")[:2]) == learner_data['courseID']:
+                course_id = course['id']
+                break
+
+        if not course_id:
+            raise CanvasClientError(
+                "Course: {course_id} not found registered in Canvas for Edx learner: {user_id}"
+                "/Canvas learner: {canvas_user_id}.".format(
+                    course_id=learner_data['courseID'],
+                    user_id=learner_data['userID'],
+                    canvas_user_id=canvas_user_id
+                ))
+
+        # Check if the course assignment already exists
+        canvas_assignments_url = '{canvas_base_url}/api/v1/courses/{course_id}/assignments'.format(
+            canvas_base_url=self.enterprise_configuration.canvas_base_url,
+            course_id=course_id
+        )
+        resp = self.session.get(canvas_assignments_url)
+        resp.raise_for_status()
+        assignments_resp = resp.json()
+        assignment_id = None
+        for assignment in assignments_resp:
+            try:
+                if assignment['integration_id'] == integration_id:
+                    assignment_id = assignment['id']
+                    break
+            except (KeyError, ValueError):
+                # This catches if the response the client got back doesn't have the expected assignment
+                # data.
```

Comment:
will leave it upto ya but I think these comments can go too.. not sure if they will help add more context to the code reader ...

Reply:
Esp. since the exception message is quite clear which is where you can put any such info!

Reply:
Sure thing!